wrappers/python/simtk/openmm/app/modeller.py

```diff
@@ -257,6 +258,120 @@ def convertWater(self, model='tip3p'):
         self.topology = newTopology
         self.positions = newPositions
 
+    def _addIons(self, forcefield, replaceableMols, ionCutoff=0.05*nanometer, positiveIon='Na+', negativeIon='Cl-', ionicStrength=0*molar, neutralize=True):
+        """Adds ions to the system by replacing certain molecules.
+
+        Parameters
+        ----------
+        forcefield : ForceField
+            the ForceField to use to determine the total charge of the system.
+        replaceableMols : dict
+            the molecules to replace by ions, as a dictionary of residue:positions
+        ionCutoff: distance=0.5*nanometer
+        positiveIon : string='Na+'
+            the type of positive ion to add.  Allowed values are 'Cs+', 'K+', 'Li+', 'Na+', and 'Rb+'
+        negativeIon : string='Cl-'
+            the type of negative ion to add.  Allowed values are 'Cl-', 'Br-', 'F-', and 'I-'. Be aware
+            that not all force fields support all ion types.
+        ionicStrength : concentration=0*molar
+            the total concentration of ions (both positive and negative) to add.  This
+            does not include ions that are added to neutralize the system.
+            Note that only monovalent ions are currently supported.
+        neutralize : bool=True
+            whether to add ions to neutralize the system
+        """
+
+        posIonElements = {'Cs+': elem.cesium, 'K+': elem.potassium,
+                          'Li+': elem.lithium, 'Na+': elem.sodium,
+                          'Rb+': elem.rubidium}
+        negIonElements = {'Cl-': elem.chlorine, 'Br-': elem.bromine,
+                          'F-': elem.fluorine, 'I-': elem.iodine}
+
+        ionPositions = []
+
+        numReplaceableMols = len(replaceableMols)
+
+        # Fetch ion elements from user input
+        if positiveIon not in posIonElements:
+            raise ValueError('Illegal value for positive ion: {}'.format(positiveIon))
+        if negativeIon not in negIonElements:
+            raise ValueError('Illegal value for negative ion: {}'.format(negativeIon))
+        positiveElement = posIonElements[positiveIon]
+        negativeElement = negIonElements[negativeIon]
+
+        # Determine the total charge of the system
+        system = forcefield.createSystem(self.topology)
+        for i in range(system.getNumForces()):
+            if isinstance(system.getForce(i), NonbondedForce):
+                nonbonded = system.getForce(i)
+                break
+        else:
+            raise ValueError('The ForceField does not specify a NonbondedForce')
+
+        totalCharge = 0.0
+        for i in range(nonbonded.getNumParticles()):
+            nb_i = nonbonded.getParticleParameters(i)
+            totalCharge += nb_i[0].value_in_unit(elementary_charge)
+        # Round up to integer
```

Comment:
It actually rounds to the nearest integer, not up.

Reply:
Why not `int(round(totalCharge))` then?

Reply:
Or why not just `int(total_charge + 0.5)`?  The `floor` call here is superfluous.

:smiley:

Personally I always round via truncating the number + 0.5.  That's a very general way of truncating to a set number of sigfigs:

Nearest 10:

```c++
(int)((N + 5) * 0.1) * 10
```

Nearest 100:
```c++
(int)((n + 50) * 0.01) * 100
```

This is equivalent to `round(N, -1)` and `round(N, -2)` in Python, respectively, but not all languages support the "places" argument for round (e.g., C).

Reply:
I don't have much preference between the different ways of doing it.  They're all pretty much equivalent.  Though I just tested them and found what I consider to be a bug in Python's `round()` function.  `round(0.5)` incorrectly returns 0 instead of 1.

Reply:
See: `https://docs.python.org/3/library/functions.html#round` - they call it a feature :)

OK, I'll just keep whatever is the norm in OpenMM, for consistency.